### PR TITLE
Topic recognition_48300588

### PR DESCRIPTION
**Pull request by Astora-Solaire**
This is a pull request contains my implementation of a GNN model, which is based on the Graph Convolutional Network (GCN) architecture proposed by [Kipf and Welling (2017)](https://arxiv.org/abs/1609.02907). The model consists of multiple graph convolutional layers that aggregate information from a node's neighbours, allowing it to capture both local and global graph structure.

Specifically it contains the files:

dataset.py: This file is about loading and processing graph data [facebook dataset](https://snap.stanford.edu/data/facebook-large-page-page-network.html).
train.py: This file trains and saves a model.
train.py: This file is about predicting the labels for nodes.
modules.py: This file is about building a Graph Convolutional Network (GCN) model.
README.md: This file includes detailed introduction and results.

Details:

Dataset size: training (70%), validation (15%), and test (15%)
hidden_channels: 64
Optimizer: SGD (lr=0.7, weight_decay=5e-4, momentum=0.95, step_size=50, gamma=0.9)
Loss: Cross-Entropy Loss
Activation Function: ReLU
Early stopping at epoch: 147
Accuracy: 0.9487

Thanks,
48300588,
Kangqi Wang

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
GNN FB Page-Page Network dataset ---- Normal Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 5 | |
| Implementation functions as intended	| 3 | 3 | 3 layer GCN, own GCN layer |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 0 | Normal Difficulty |
| Section IV : Max mark 15 from 20		|   | **15**   | |

**TSNE / UMAP:** 
Good.
- Discussion: Good, with a bit more depth useful.

**Suggestions:**
Comprehensive.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. Great job.</td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Some/Adequate Progressive commits. Somewhat minimalistic.</td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-6</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (README). Please update for merge, doesn't affect grade.
